src/Note.cpp

```diff
@@ -0,0 +1,378 @@
+/*************************************************************************/
+/*  Xania (M)ulti(U)ser(D)ungeon server source code                      */
+/*  (C) 1995-2000 Xania Development Team                                 */
+/*  See the header to file: merc.h for original code copyrights          */
+/*                                                                       */
+/*  Note.cpp: notes system                                               */
+/*                                                                       */
+/*************************************************************************/
+
+#include "Note.hpp"
+#include "DescriptorList.hpp"
+#include "TimeInfoData.hpp"
+#include "db.h"
+#include "merc.h"
+#include "string_utils.hpp"
+
+#include <fmt/format.h>
+#include <range/v3/algorithm/find_if.hpp>
+#include <range/v3/view/filter.hpp>
+
+#include <cctype>
+#include <cstdio>
+#include <functional>
+
+static const char *NOTE_FILE = "notes.txt";
+
+bool Note::is_to(const Char &ch) const {
+    if (matches(ch.name, sender_)) {
+        return true;
+    }
+    if (is_name("all", to_list_)) {
+        return true;
+    }
+    if (ch.is_immortal() && is_name("immortal", to_list_)) {
+        return true;
+    }
+    if (is_name(ch.name, to_list_)) {
+        return true;
+    }
+    return false;
+}
+
+bool Note::sent_by(const Char &ch) const { return matches(ch.name, sender_); }
+
+void Note::save(FILE *file) const {
+    fmt::print(file, "Sender {}~\n", sender_);
+    fmt::print(file, "Date {}~\n", date_);
+    fmt::print(file, "Stamp {}\n", Clock::to_time_t(date_stamp_));
+    fmt::print(file, "To {}~\n", to_list_);
+    fmt::print(file, "Subject {}~\n", subject_);
+    fmt::print(file, "Text\n{}~\n", text_);
+}
+
+void Notes::erase(const Note &note) {
+    auto it = ranges::find_if(notes_, [&](const Note &data) { return &data == &note; });
+    if (it != notes_.end())
+        notes_.erase(it);
+}
+Note &Notes::add(Note note) { return notes_.emplace_back(std::move(note)); }
```

Comment:
Q: The arg is not passing a reference or a pointer here,  which makes me think it would be a copy by value, but would passing as a reference be any better?

Reply:
TLDR; this nets out as more efficient in some cases if you know you're going to be copying one of your parameters. The clang linter recommends this approach in many cases (so I've gotten used to doing it as a habit).

---

This is a fun one, and mainly goes by the theory that if you're going to be taking of copy of something (as we are when we put the note into the vector), then you should make that copy explicit. In this instance the "copy" happens in the parameter name.

The interesting thing about this is that by _moving_ from the parameter you can get nearly perfect forwarding/construction behaviour.

Take for example some code like this, which takes a copy of a string:

```
using namespace std; // lazy
struct Thing {
  string s_;
  void set_s(string s) { s_ = move(s); }
};
```

Now consider what happens when we do something like `set_s("moo");`:

* the `const char*` gets converted into a temporary `string` via `string`'s constructor that takes a const char*. That does allocations^ etc and copies the "moo" into the new temporary `string` object`
* that temporary is used to construct the parameter value `s` in the `set_s`. The argument to the constructor of this `s` is a temporary - so it's of type `string &&`. The `string&&` constructor can "steal" the contents of the temporary `string&&` which makes it super cheap (no allocations, just swapping pointers)
* the `s_ = std::move()` runs. The `s_`'s `operator=(string &&)` is called (as the `move` turns the `string` into an r-value). This _also_ does the steal-the-innards trick.
* All the destructors for the temporaries (the original "moo" object and the pass-by-copy param `s`) run. They have no work to do as they've been "stolen" from.

Short version: exactly one "real" string construction is done, and then there's a bit of pointer movement, but nothing more expensive.

If, however, we change to a `const string &` in `set_s`:
* the `const char *` gets converted into a temporary `string` as before.
* it is passed by reference (super cheap) to `set_s()` - no work done here at all.
* In the `s_ =  s;` we copy the string object. This is another allocation and copy, as we can't "steal" anything from anyone
* we then destroy the temporary `string`

This means we do the allocation twice, copy the string data twice, and do a deallocation once. More expensive! https://godbolt.org/z/n9vqx9

---

^ ok so actually no real allocations are done in this case as there's something called the "small string optimization" (SSO) kicking in; the string object holds small strings _by value_ inside the object. But,,,the work is still there. If you change the string to something longer you see the work https://godbolt.org/z/n9vqx9 , but it's more complex :) The expense is in the `_M_create` method you'll see in the asm output.